ada/language/ast.py

```diff
@@ -485,9 +485,47 @@ def parent_basic_decl():
         """
         Return the parent basic decl for this node, if applicable, null
         otherwise.
+
+        .. note:: If the parent BasicDecl of the given node is a generic
+            declaration, this call will return the instantiation from which
+            the node was retrieved instead, if any.
         """
         return Entity.semantic_parent.then(
-            lambda sp: sp.cast(T.BasicDecl)._or(sp.parent_basic_decl)
+            lambda sp: If(
+                sp.is_a(GenericPackageInternal) | sp.is_a(GenericSubpInternal),
+                Let(
+                    lambda
+                    i=sp.info.rebindings._.new_env.env_node.cast(BasicDecl):
```

Comment:
Could you please give this variable a more meaningful name?

Reply:
Done